File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/federation/FederationExpect.scala
package ch.epfl.pop.model.network.method.message.data.federation

import ch.epfl.pop.json.MessageDataProtocol.FederationExpectFormat
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.Message
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Hash, PublicKey}
import spray.json.*

final case class FederationExpect(
    laoId: Hash,
    serverAddress: String,
    publicKey: PublicKey,
    challenge: Message
) extends MessageData {
  override val _object: ObjectType = ObjectType.federation
  override val action: ActionType = ActionType.expect
}

object FederationExpect extends Parsable {
  def apply(laoId: Hash, serverAddress: String, publicKey: PublicKey, challenge: Message): FederationExpect = {
    new FederationExpect(laoId, serverAddress, publicKey, challenge)
  }

  override def buildFromJson(payload: String): FederationExpect = payload.parseJson.asJsObject.convertTo[FederationExpect]
}

File: be1-go/message/messagedata/federation_expect.go
package messagedata

import "popstellar/message/query/method/message"

// FederationExpect defines a message data
type FederationExpect struct {
	Object string `json:"object"`
	Action string `json:"action"`

	LaoId         string          `json:"lao_id"`
	ServerAddress string          `json:"server_address"`
	PublicKey     string          `json:"public_key"`
	ChallengeMsg  message.Message `json:"challenge"`
}

// GetObject implements MessageData
func (FederationExpect) GetObject() string {
	return FederationObject
}

// GetAction implements MessageData
func (FederationExpect) GetAction() string {
	return FederationActionExpect
}

// NewEmpty implements MessageData
func (FederationExpect) NewEmpty() MessageData {
	return &FederationExpect{}
}
